resolver.go

```diff
@@ -145,6 +145,14 @@ func (r *resolver) Start() error {
 		return fmt.Errorf("setting up IP table rules failed: %v", err)
 	}
 
+	// If the container goes down quickly after start resolver.Stop() can
+	// get called before Start() completes. Check if the socket is still
+	// valid before Activating the DNS server. This avoids a panic in dns
+	// libarary code because of missing nil check for interface's data.
+	// github.com/docker/docker/issues/28112
+	if r.conn == nil {
```

Comment:
`Yes it is, the stop() may be called after the start() started the udp listener but before it started the tcp listener.` 

Both udp & tcp sockets are created in `SetupFunc` before `Start()`. Unless both are created resolver.err will be non-nil which is already checked. So its not possible to have `resolver.conn != nil && resolver.tcpListen == nil`


`I do not think it is the library responsibility if we nil the net.PacketConn after initializing its server structure.` 

We are not setting `net.PacketConn`  to `nil` after creating `dns.Server`; rather `net.PacketConn` is set to `r.conn` which happens to be nil. Since `net.PacketConn` is an interface type the library should check the interface value before accessing it. ie: the following is not safe

https://github.com/miekg/dns/blob/3f1f7c8ec9ead89493df11f2c3d8bec353a2c2c0/server.go#L392